### PR TITLE
12019: user can create new contact record associated with project on project editors list

### PR DESCRIPTION
**Pull request by trbmcginnis**
Allow user to create new "project editors" on the project page. 

**Overall Behavior**
- When a user enters first name, last name, and email address into the project editor form, they are prompted with a confirmation modal and allowed to save this new project editor. 
- If a contact that matches that email address is already created, then the only POST that occurs is on the project applicant entity associated with the current project. 
- If a contact does NOT exist that matches the email address entered, then a POST will occur on both the contact entity and the project applicant entity

**Server Behavior**
- While we are only creating a record and saving it for project applicants on the frontend, we also POST a new contact in the server when we POST on the `project-applicant.controller`

**ToDo**
- separate out project editor form into own component
- update design on project editor form and associated confirmation modal
- incorporate activation/deactivation scenarios for project applicant
- implement form validation / update form with UI components

Closes [AB#12019](https://dcp-paperless.visualstudio.com/d36fd830-9029-4b77-b0c4-b0df2012eb98/_workitems/edit/12019)

### CODE REVIEW

**Comment by godfreyyeung** [comment]
> Allow user to create new "project editors" on the project page.
> 
> **Overall Behavior**
> 
> * When a user enters first name, last name, and email address into the project editor form, they are prompted with a confirmation modal and allowed to save this new project editor.
> * If a contact that matches that email address is already created, then the only POST that occurs is on the project applicant entity associated with the current project.
> * If a contact does NOT exist that matches the email address entered, then a POST will occur on both the contact entity and the project applicant entity
> 
> **Server Behavior**
> 
> * While we are only creating a record and saving it for project applicants on the frontend, we also POST a new contact in the server when we POST on the `project-applicant.controller`
> 
> **ToDo**
> 
> * separate out project editor form into own component
> * update design on project editor form and associated confirmation modal
> * incorporate activation/deactivation scenarios for project applicant
> * implement form validation / update form with UI components
> 
> Closes [AB#12019](https://dcp-paperless.visualstudio.com/d36fd830-9029-4b77-b0c4-b0df2012eb98/_workitems/edit/12019)

Great writeup!

File: server/src/projects/project-applicants/project-applicant.controller.ts [hunk 1]
```diff
@@ -0,0 +1,85 @@
+import { Controller, Patch, Body, Param, Post, UseInterceptors, UseGuards, UsePipes, Delete, HttpException, HttpStatus } from '@nestjs/common';
```
Comment (godfreyyeung) [inline server/src/projects/project-applicants/project-applicant.controller.ts:1]: Looks like `Patch` and `Param` are imported but not used -- we should delete them

File: server/src/contact/contact.controller.ts [hunk 2]
```diff
@@ -1,6 +1,9 @@
-import { Controller, Session, Get, UseInterceptors, HttpException, HttpStatus } from '@nestjs/common';
+import { Controller, Session, Get, UseInterceptors, HttpException, HttpStatus, Post, Body } from '@nestjs/common';
```
Comment (godfreyyeung) [inline server/src/contact/contact.controller.ts:1]: Looks like we can delete the unused `Post`, `Body`, `import { pick } from 'underscore';` and also `import { CONTACT_ATTRS } from './contacts.attrs';`

File: server/src/projects/project-applicants/project-applicant.controller.ts [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+@UseGuards(AuthenticateGuard)
+@UsePipes(JsonApiDeserializePipe)
+@Controller('project-applicants')
+export class ProjectApplicantController {
+  constructor(private readonly crmService: CrmService) {}
+
+  @Post('/')
+  async create(@Body() body) {
+    const allowedAttrs = pick(body, PROJECTAPPLICANT_ATTRS);
+    const email = body.emailaddress;
```
Comment (godfreyyeung) [inline server/src/projects/project-applicants/project-applicant.controller.ts:34]: optional alternate style: 
```
const {
  emailaddress: email,
  project: projectId,
  dcp_name: fullName
} = body;

File: server/src/projects/project-applicants/project-applicant.controller.ts [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+    const projectId = body.project;
+    const fullName = body.dcp_name;
+
+    let contactId;
+
+    // check if contact already exists for this email address
+    const { records } = await this.crmService.get('contacts', `
+      $select=${CONTACT_ATTRS.join(',')}
+      &$filter=startswith(emailaddress1, '${email}')
+        and statuscode eq ${ACTIVE_CODE}
```
Comment (godfreyyeung) [inline server/src/projects/project-applicants/project-applicant.controller.ts:44]: Should we find ALL contacts first, and then reactivate a contact if it already exists? 
Here, the code might create a contact that already exists but is just deactivated 

No problem if you wanna do it in a separate PR, though

File: server/src/projects/project-applicants/project-applicant.controller.ts [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+  constructor(private readonly crmService: CrmService) {}
+
+  @Post('/')
+  async create(@Body() body) {
+    const allowedAttrs = pick(body, PROJECTAPPLICANT_ATTRS);
+    const email = body.emailaddress;
+    const projectId = body.project;
+    const fullName = body.dcp_name;
+
+    let contactId;
```
Comment (godfreyyeung) [inline server/src/projects/project-applicants/project-applicant.controller.ts:38]: can you move contactId down to maybe line 47? It feels a bit far from where it's actually used

File: server/src/projects/project-applicants/project-applicant.controller.ts [hunk 6]
```diff
@@ -0,0 +1,85 @@
...
+
+    let newContact = {
+      contactid: null,
+    };
+
+    if (records.length > 0) {
+      contactId = records[0].contactid;
+    } else {
+      const regex = /^([\w\-]+)/g;
+      const firstName = fullName.match(regex)[0];
```
Comment (godfreyyeung) [inline server/src/projects/project-applicants/project-applicant.controller.ts:56]: Is this grabbing the first token and using it as a first name?

I wonder how we should handle names if the first name has multiple parts, separated by a space
Like
`Sandra Day O'Connor`

File: server/src/projects/project-applicants/project-applicant.controller.ts [hunk 7]
```diff
@@ -0,0 +1,85 @@
...
+      HttpStatus.UNPROCESSABLE_ENTITY,
+    );
+
+    const newProjectApplicant = await this.crmService.create('dcp_projectapplicants', {
+      ...allowedAttrs,
+
+      // Dy365 syntax for associating a newly-created record
+      // with an existing record.
+      // see: https://docs.microsoft.com/en-us/powerapps/developer/common-data-service/webapi/create-entity-web-api#associate-entity-records-on-create
+      ...(body.project ? { 'dcp_Project@odata.bind': `/dcp_projects(${projectId})` } : {}),
```
Comment (godfreyyeung) [inline server/src/projects/project-applicants/project-applicant.controller.ts:77]: why not use `projectId`  here instead of `body.project`?

File: server/src/projects/project-applicants/project-applicant.controller.ts [hunk 8]
```diff
@@ -0,0 +1,85 @@
...
+      const lastName = fullName.replace(`${firstName} `, '');
+      newContact = await this.crmService.create(`contacts`, {
+        'emailaddress1': email,
+        'firstname': firstName,
+        'lastname': lastName,
+      });
+      contactId = newContact.contactid;
+    }
+
+    if (!body.emailaddress || !body.project) throw new HttpException(
```
Comment (godfreyyeung) [inline server/src/projects/project-applicants/project-applicant.controller.ts:66]: why not use `!email || projectId`  here?

Comment (godfreyyeung) [reply server/src/projects/project-applicants/project-applicant.controller.ts:66]: Also, can this check be moved to the top of the function? Like to line 38, since it looks like the contact creation code depends on `email`

File: server/src/projects/project-applicants/project-applicant.controller.ts [hunk 9]
```diff
@@ -0,0 +1,85 @@
...
+      'Missing email address or project id',
```
Comment (godfreyyeung) [inline server/src/projects/project-applicants/project-applicant.controller.ts:67]: Can you add `title`, `code` and `detail` properties here? 
like 
```
{
  code: 'ADD_APPLICANT_ERROR'
  title: 'Add Applicant failed',
  detail: 'Cannot add applicant because missing ....',
}
```

File: server/src/projects/project-applicants/project-applicant.controller.ts (hunk 6 above)
Comment (godfreyyeung) [reply server/src/projects/project-applicants/project-applicant.controller.ts:56]: Maybe we should instead grab the last token and use it as the last name?

File: server/src/projects/project-applicants/project-applicant.controller.ts (hunk 3 above)
Comment (godfreyyeung) [reply server/src/projects/project-applicants/project-applicant.controller.ts:34]: Wait.. here I think we should use `allowedAttrs` instead of `body`

File: server/src/projects/project-applicants/project-applicant.controller.ts (hunk 6 above)
Comment (trbmcginnis) [reply server/src/projects/project-applicants/project-applicant.controller.ts:56]: Yeah I was thinking about this too -- it sucks that projectApplicants have one field for the name and contacts has two. I know we concatenate the names from contact --> projectApplicant, but it makes it difficult to go the other way

File: server/src/projects/project-applicants/project-applicant.controller.ts (hunk 4 above)
Comment (trbmcginnis) [reply server/src/projects/project-applicants/project-applicant.controller.ts:44]: Ooh are `contacts` also able to be activated/deactivated? I thought it was only projectApplicants

File: server/src/projects/project-applicants/project-applicant.controller.ts [hunk 10]
```diff
@@ -0,0 +1,94 @@
...
+export class ProjectApplicantController {
+  constructor(private readonly crmService: CrmService) {}
+
+  @Post('/')
+  async create(@Body() body) {
+    const allowedAttrs = pick(body, PROJECTAPPLICANT_ATTRS);
+    const {
+      emailaddress: email,
+      project: projectId,
+    } = body;
```
Comment (godfreyyeung) [inline server/src/projects/project-applicants/project-applicant.controller.ts:41]: I think you can do `... } = allowedAttrs` here?

File: server/src/projects/project-applicants/project-applicant.controller.ts [hunk 11]
```diff
@@ -0,0 +1,94 @@
...
+
+    let newContact = {
+      contactid: null,
+    };
+
+    if (records.length > 0) {
+      contactId = records[0].contactid;
+      // if contact record is deactivated, reactive it
+      if (records[0].statuscode === INACTIVE_CONTACT_STATUSCODE && records[0].statecode === INACTIVE_CONTACT_STATECODE) {
+        this.crmService.update('contacts', contactId, {
```
Comment (godfreyyeung) [inline server/src/projects/project-applicants/project-applicant.controller.ts:68]: Nice!

File: server/src/projects/project-applicants/project-applicant.controller.ts [hunk 12]
```diff
@@ -0,0 +1,94 @@
...
+          statuscode: ACTIVE_CONTACT_STATUSCODE,
+          statecode: ACTIVE_CONTACT_STATECODE,
+        });
+      }
+    } else {
+      newContact = await this.crmService.create(`contacts`, {
+        'emailaddress1': email,
```
Comment (godfreyyeung) [inline server/src/projects/project-applicants/project-applicant.controller.ts:75]: Nice! No dealing with first and last name anymore

File: server/src/projects/project-applicants/project-applicant.controller.ts [hunk 13]
```diff
@@ -0,0 +1,94 @@
...
+
+  @Post('/')
+  async create(@Body() body) {
+    const allowedAttrs = pick(body, PROJECTAPPLICANT_ATTRS);
+    const {
+      emailaddress: email,
+      project: projectId,
+    } = body;
+
+    if (!email || !projectId) {
```
Comment (godfreyyeung) [inline server/src/projects/project-applicants/project-applicant.controller.ts:43]: Awesome, I like the early check and exception
